### DISCUSSION TITLE
How to build Argontlunar 2 LV2 plugin on Debian

### DISCUSSION THREAD

**Discussion by ChrisDAT20** [DISTRHO/DISTRHO-Ports#127] [General]
I want to build the amd64 LV version of Argontlunar 2. I try to figure out what needs to be downloaded from DISTRHO and how I can build the plugin.
Is there a page which describes those steps?

What I tried so far:

git clone https://github.com/DISTRHO/DISTRHO-Ports.git
copy contents of directory DISTRHO-Ports to /usr/src
run sudo meson setup build --buildtype release

Following output is the result:

```
user@laptop:/usr/src/distrho$ sudo meson setup build --buildtype release
The Meson build system
Version: 1.2.3
Source dir: /usr/src/distrho
Build dir: /usr/src/distrho/build
Build type: native build
Project name: DISTRHO-Ports
Project version: 1
C compiler for the host machine: cc (gcc 13.2.0 "cc (Debian 13.2.0-10) 13.2.0")
C linker for the host machine: cc ld.bfd 2.41.90.20240122
C++ compiler for the host machine: c++ (gcc 13.2.0 "c++ (Debian 13.2.0-10) 13.2.0")
C++ linker for the host machine: c++ ld.bfd 2.41.90.20240122
Host machine cpu family: x86_64
Host machine cpu: x86_64
Library dl found: YES
Library rt found: YES
Found pkg-config: /usr/bin/pkg-config (1.8.1)
Run-time dependency fftw3f found: YES 3.3.10
Run-time dependency alsa found: YES 1.2.10
Run-time dependency freetype2 found: YES 26.1.20
Run-time dependency x11 found: YES 1.8.7
Run-time dependency xext found: YES 1.3.4
Library dl found: YES

libs/drowaudio/meson.build:11:8: ERROR: Include dir ../juce5/source/modules does not exist.
```

**Comment by falkTX** [comment]
you need to either clone recursively or init + update the git submodules after cloning

**Reply by ChrisDAT20** [reply]
Yes, I already figured out that there are linked branches in the lib folders. git clone --recursive fetched them all.

**Reply by ChrisDAT20** [reply]
How can I reduce the stuff which will be installed to the system, as I only want the Argotlunar LV2 plugin?

**Reply by ChrisDAT20** [reply]
If I am not wrong, it is possible with meson to add parameters to the build command like `build-lv2` and `plugins`, the first a boolean and the later a list. But I need to add the extra module to get a hold onto Argontlunar.

**Comment by ChrisDAT20** [comment]
I proceeded with DISTRHO-Ports-Extra with git clone --recursive, copied the contents to /usr/src/distrho-extra.
But when I execute `./scripts/premake-update.sh linux`, it tells me that `./scripts/premake-update.sh` could not be found.
That script exists only in the `old` sub folder of `/usr/src/distrho/scripts`. Can I just copy it to the expected location?

**Comment by ChrisDAT20** [comment]
Somehow DISTRHO-Ports-Extra does not match the build env from DISTRHO-Ports. DISTRHO-Ports-Extra uses make, and refers to scripts in `old`. And make traverses into ports/argontlunar/LV2, and there is a LUA script. And that LUA script refers another LUA script, which has been moved to `old` as well. And the move to `old` is some 4 years ago. So it seems nobody build Argontlunar plugins since then? Or am I completely wrong and missed how to "merge" the extra stuff into the DISTRHO source tree?

**Reply by ChrisDAT20** [reply]
Copied the needed scripts back into the scripts folder, and needed to add some paths to the includes in `premake.lua`. I assumed juce5 is correct. It compiles. But there are issues when linking:
```
Linking argotlunar2.lv2/argotlunar2
/usr/bin/ld: cannot find -ljuce-legacy: No such file or directory
/usr/bin/ld: cannot find -ljuce-legacy: No such file or directory
collect2: error: ld returned 1 exit status
make[3]: *** [argotlunar2.lv2/argotlunar2.make:85: ../../../bin/lv2/argotlunar2.lv2/argotlunar2.so] Error 1
make[2]: *** [Makefile:22: argotlunar2.lv2/argotlunar2] Error 2
make[2]: Leaving directory '/usr/src/distrho-extra/ports/argotlunar/LV2'
make[1]: *** [Makefile:13: lv2] Error 2
make[1]: Leaving directory '/usr/src/distrho-extra/ports'
make: *** [Makefile:61: lv2] Error 2
```
`juce-legacy` is referred in `make-project.lua`. I could not find where I can get that one.

**Reply by falkTX** [reply]
juce-legacy is juce5

**Comment by falkTX** [comment]
those extra plugins likely should be just ported back into the main project. I put them in the extra repo because at the time they were still actively maintained. but that is not the case anymore, so no need for the separation

**Reply by ChrisDAT20** [reply]
there was some recent activities, at least it was ported to JUCE7 according to its repo:
https://github.com/mourednik/argotlunar/tree/2.1.0

**Reply by ChrisDAT20** [reply]
I tried to port argotlunar into DISTRHO-Ports, but without success so far. I added a git submodule to /usr/src/distrho/ports-juce7 from branch 2.1.0 (https://github.com/mourednik/argotlunar/tree/2.1.0) and copied over the meson.build files which I took from a juce6.1 plugin. I adapted them to what I thought should work. Both meson.build files are attached.

![image](https://github.com/DISTRHO/DISTRHO-Ports/assets/92817778/51901b8b-359b-494f-94a1-c898ae39e42c)

The build is configured as follows:
```
Build Options: -Dbuildtype=release -Dbuild-vst2=false -Dbuild-vst3=false -Dplugins=argotlunar -Dbuild-juce7-only=true
``` 
And the meson-log.txt contains the following output:
```
Run-time dependency xext found: YES 1.3.4
Library dl found: YES
Message: Building plugin ->  argotlunar
Message: In argotlunar!
Build targets in project: 6
NOTICE: Future-deprecated features used:
 * 0.56.0: {'meson.source_root'}

DISTRHO-Ports 1

  User defined options
    buildtype       : release
    build-juce7-only: true
    build-vst2      : false
    build-vst3      : false
    plugins         : argotlunar

Found ninja-1.11.1 at /usr/bin/ninja
Failed to guess install tag for /usr/local/lib/lv2/Argotlunar.lv2
Failed to guess install tag for /usr/local/lib/lv2/Argotlunar.lv2
```

[_ports-juce7_argotlunar_meson.build.zip](https://github.com/DISTRHO/DISTRHO-Ports/files/14102894/_ports-juce7_argotlunar_meson.build.zip)
[_ports-juce7_meson.build.zip](https://github.com/DISTRHO/DISTRHO-Ports/files/14102895/_ports-juce7_meson.build.zip)

**Reply by ChrisDAT20** [reply]
[stdout-argotlunar-build.txt](https://github.com/DISTRHO/DISTRHO-Ports/files/14103236/stdout-argotlunar-build.txt)
Tried the same with juce5 based argotlunar, with same results:

```
Command line: `cc /usr/src/distrho/build/meson-private/tmpn80pzj0y/testfile.c -o /usr/src/distrho/build/meson-private/tmpn80pzj0y/output.exe -D_FILE_OFFSET_BITS=64 -O0 -Wl,--start-group -lrt -Wl,--end-group -Wl,--allow-shlib-undefined` -> 0
Library rt found: YES
Dependency fftw3f found: YES 3.3.10 (cached)
Dependency alsa found: YES 1.2.10 (cached)
Dependency freetype2 found: YES 26.1.20 (cached)
Dependency x11 found: YES 1.8.7 (cached)
Dependency xext found: YES 1.3.4 (cached)
Library dl found: YES
Dependency x11 found: YES 1.8.7 (cached)
Dependency xext found: YES 1.3.4 (cached)
Message: In argotlunar!
Build targets in project: 8
NOTICE: Future-deprecated features used:
 * 0.56.0: {'meson.source_root'}

DISTRHO-Ports 1

  User defined options
    buildtype       : release
    build-juce5-only: true
    build-juce7-only: false
    build-vst2      : false
    build-vst3      : false
    plugins         : argotlunar

Found ninja-1.11.1 at /usr/bin/ninja
Running compile:
Working directory:  /usr/src/distrho/build/meson-private/tmpqjbky1r4
Code:
 
-----------
Command line: `c++ /usr/src/distrho/build/meson-private/tmpqjbky1r4/testfile.cpp -o /usr/src/distrho/build/meson-private/tmpqjbky1r4/output.obj -c -D_FILE_OFFSET_BITS=64 -O0 -fpermissive --print-search-dirs` -> 0
stdout:
install: /usr/lib/gcc/x86_64-linux-gnu/13/
programs: =/usr/libexec/gcc/x86_64-linux-gnu/13/:/usr/libexec/gcc/x86_64-linux-gnu/13/:/usr/libexec/gcc/x86_64-linux-gnu/:/usr/lib/gcc/x86_64-linux-gnu/13/:/usr/lib/gcc/x86_64-linux-gnu/:/usr/lib/gcc/x86_64-linux-gnu/13/../../../../x86_64-linux-gnu/bin/x86_64-linux-gnu/13/:/usr/lib/gcc/x86_64-linux-gnu/13/../../../../x86_64-linux-gnu/bin/x86_64-linux-gnu/:/usr/lib/gcc/x86_64-linux-gnu/13/../../../../x86_64-linux-gnu/bin/
libraries: =/usr/lib/gcc/x86_64-linux-gnu/13/:/usr/lib/gcc/x86_64-linux-gnu/13/../../../../x86_64-linux-gnu/lib/x86_64-linux-gnu/13/:/usr/lib/gcc/x86_64-linux-gnu/13/../../../../x86_64-linux-gnu/lib/x86_64-linux-gnu/:/usr/lib/gcc/x86_64-linux-gnu/13/../../../../x86_64-linux-gnu/lib/../lib/:/usr/lib/gcc/x86_64-linux-gnu/13/../../../x86_64-linux-gnu/13/:/usr/lib/gcc/x86_64-linux-gnu/13/../../../x86_64-linux-gnu/:/usr/lib/gcc/x86_64-linux-gnu/13/../../../../lib/:/lib/x86_64-linux-gnu/13/:/lib/x86_64-linux-gnu/:/lib/../lib/:/usr/lib/x86_64-linux-gnu/13/:/usr/lib/x86_64-linux-gnu/:/usr/lib/../lib/:/usr/lib/gcc/x86_64-linux-gnu/13/../../../../x86_64-linux-gnu/lib/:/usr/lib/gcc/x86_64-linux-gnu/13/../../../:/lib/:/usr/lib/
-----------
Failed to guess install tag for /usr/local/lib/lv2/Argotlunar.lv2
Failed to guess install tag for /usr/local/lib/lv2/Argotlunar.lv2
```
